mlflow/sklearn/__init__.py

```diff
@@ -455,6 +455,66 @@ def load_model(model_uri):
     )
 
 
+from collections import defaultdict
+
+
+class _AutologTrainingStatus:
+    def __init__(self, last_mlflow_run_id=None, model_id=None):
+        self.last_mlflow_run_id = last_mlflow_run_id
+        self.model_id = model_id
+        self.log_test_metric_count_map = defaultdict(lambda: 0)
+        self.predict_result_instance_ids = []
+        self.in_fit_call_scope = False
+        self.in_scorer_call_scope = False
+
+    def register_predition_result(self, model, pred_y):
+        if id(model) == self.model_id:
+            self.predict_result_instance_ids.append(id(pred_y))
+
+    def get_log_test_metric_index(self, metric_name):
+        metric_index = self.log_test_metric_count_map[metric_name]
+        self.log_test_metric_count_map[metric_name] += 1
+        return metric_index
+
+    def should_log_test_metric_for(self, metric_api_call_arg_list):
+        # Note: some metric API the arguments is not `y_true`, `y_pred`
+        #  e.g.
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.roc_auc_score.html#sklearn.metrics.roc_auc_score
+        #    https://scikit-learn.org/stable/modules/generated/sklearn.metrics.silhouette_score.html#sklearn.metrics.silhouette_score
+        # So here check whether predict_result_instance_ids includes instance id from `predict_result_instance_ids`
+        active_run = mlflow.active_run()
+        return active_run and active_run.info.run_id == self.last_mlflow_run_id \
+            and any(id(arg) in self.predict_result_instance_ids
+                    for arg in metric_api_call_arg_list)
+
+
+_autolog_training_status = _AutologTrainingStatus()
+
+
+class ResumeAutologRun:
+    def __init__(self):
+        self.managed_run = None
+        self.run = None
+        cur_run = mlflow.active_run()
+        if cur_run is not None:
+            # Reuse current active run.
+            self.run = cur_run
+        else:
+            if _autolog_training_status.last_mlflow_run_id is not None:
+                self.run = mlflow.start_run(
+                    run_id=_autolog_training_status.last_mlflow_run_id
+                )
+                self.managed_run = self.run
+
+    def __enter__(self):
+        return self.run
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.managed_run:
+            self.managed_run.__exit__(exc_type, exc_val, exc_tb)
```

Comment:
In mlflow user may write code in two way:

(1)
```
mlflow.autolog()
with mlflow.start_run():  # start a run, the run is active within the `with` scope.
   model.fit(..)   # autolog into the active run
   ...
   a_metric_api(...)  # autolog metric into the same active run
```

(2)
```
mlflow.autolog()
model.fit(..)   # mlflow will start a run, autolog into the run, then end the run
...
a_metric_api(...)  # resume the run created in above `fit` call, and log the metric
```

The `ResumeAutologRun` is designed to address the two cases in a uniform way.